Add Script alias

marklundin opened the pull request:
Having some discussions with @willeastcott, with the new ESM Script, we want to keep the default script boilerplate as clean as possible. This PR exports an alias for `ScriptType` named `Script`

```javascript
import { Script } from 'playcanvas';

class Rotator extends Script {
  initialize() {
    ...
  }
}
```

UPDATE:
ScriptType has been added to the deprecation.js file
Improved dicunmentation for Script to include ESM example


I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

kpal81xd commented:
What was the reason for this?

mvaligursky reviewed (dismissed):
Looks good.
Should we rename `ScriptType` to `Script` in that module, and export it as `ScriptType` for compatibility?

willeastcott commented:
> Should we rename ScriptType to Script in that module, and export it as ScriptType for compatibility?

Yeah, because otherwise, the API ref will continue to show `ScriptType`. The aliasing should be done in the backwards compatibility module.

marklundin commented:
@willeastcott are we happy to simply rename the class and keep the `script-type.js` filename?

mvaligursky commented:
Rename filename as well, we keep those in sync.

willeastcott commented:
Maybe do:

* `script.js` -> `script-create.js` (or `script-register.js`)
* `script-type.js` -> `script.js`

willeastcott commented:
BTW, we should consider deprecating and hiding `registerScript`.

mvaligursky commented:
We also need to consider the impact in the Editor. Will the Editor be able to still use at least few previous engine versions? Do we need to add some type checking to it or similar? Maybe it all just works, but we need to check.

marklundin commented:
> * `script.js` -> `script-create.js` (or `script-register.js`)
> * `script-type.js` -> `script.js`

Makes sense, although strictly speaking this would be a breaking change right?

> We also need to consider the impact in the Editor. Will the Editor be able to still use at least few previous engine versions? Do we need to add some type checking to it or similar? Maybe it all just works, but we need to check.

Yeah it should be fine, editor only uses ScriptType when parsing and `pc.Script === pc.ScriptType`

marklundin commented:
Actually maybe not a breaking change since we haven't yet [published the ESM build ](https://www.npmjs.com/package/playcanvas?activeTab=code). Mainly thinking about `import { Script } from 'playcanvas/framework/script/script-type.js'

marklundin commented:
This has not been updated

- Renamed script.js -> script-create.js
- Renamed script-type.js -> script.js
- Updated jsdoc references of {ScriptType} => {Script}

Verified with type docs...
<img width="951" alt="image" src="https://github.com/playcanvas/engine/assets/430764/6c08c208-9fc5-426b-ab35-a29a18f63d7a">

*Note that now ScriptType does not exist, it does not generate a corresponding url. It would be great to retain the original, to indicated to users that it has been renamed

willeastcott commented on src/framework/script/script.js:
```diff
@@ -1,156 +1,393 @@
-/* eslint-enable jsdoc/check-examples */
 
-export { createScript, registerScript, getReservedScriptNames };
+export { Script, Script as ScriptType };
```
I would handle the aliasing in `deprecated.js`.

marklundin replied:
You mean adding `export { Script as ScriptType }` to deprecated.js?

willeastcott replied:
I was thinking something like: https://github.com/playcanvas/engine/blob/c5177982f9c9064df1d5a881a513a302054d60ae/src/deprecated/deprecated.js#L725

Maksims commented on src/framework/script/script.js:
````diff
@@ -7,11 +7,20 @@ import { ScriptAttributes } from './script-attributes.js';
+ * ```javascript
+ * class Rotator extends Script {
+ *    update() {
+ *      this.entity.rotate(0, 0.1, 0);
````
Should be 4 spaces instead of 2.

willeastcott commented on src/framework/script/script-create.js:
```diff
@@ -0,0 +1,156 @@
+import { AppBase } from '../app-base.js';
+
+import { ScriptAttributes } from './script-attributes.js';
+import { ScriptType } from './script.js';
```
`ScriptType` -> `Script`

willeastcott commented on src/framework/script/script-create.js:
```diff
@@ -0,0 +1,156 @@
+
+    const scriptType = function (args) {
+        EventHandler.prototype.initEventHandler.call(this);
+        ScriptType.prototype.initScriptType.call(this, args);
```
`ScriptType` -> `Script`

marklundin commented:
jsdocs update

<img width="986" alt="image" src="https://github.com/playcanvas/engine/assets/430764/46e6f766-41a1-48a9-a56d-09d4d24458f7">

marklundin commented:
Any more updates on this @willeastcott? Keen to get this merged
